Model/VomsMember.php

```diff
@@ -256,4 +306,133 @@ public function getCertificates($username, $co_id) {
+        'Accept'        => 'application/json',
+        'X-API-Key'     =>  $this->authKey
+      )
+    );
+
+    $results = $this->Http->$action(
+      $this->baseUrl . $this->pathURL,
+      ($action == 'get' ? $data : json_encode($data)),
+      $options);
+
+    if($results->code != 200) {
+      // This is probably an RDF blob, which is slightly annoying to parse.
+      // Rather than do it properly since we don't parse RDF anywhere else,
+      // we return a generic error.
+      throw new RuntimeException(_txt('er.vo_members.http.failed', array($results->code)));
+    }
+
+    return json_decode($results->body, true);
+  }
+
+
+  /**
+   * Parse the data into what my database expects to save
+   *
+   * @param $data     JSON Decoded data
+   * @return array    Create list of values ready for saveMany method
+   */
+  public function parseOpsResponse($data) {
+      if(empty($data)) {
+        return array();
+      }
+
+      $values = array();
+      foreach ($data as $idx => $items) {
+        foreach ($items['row'] as $item) {
+          $column = key($item);
+          if($column == 'fqans') {
+            $values[$idx][ $this->mapper[$column] ] = serialize(json_encode($item[$column][0])) ?? '';
+          } else {
+            $values[$idx][ $this->mapper[$column] ] = $item[$column][0] ?? '';
+          }
+        }
+      }
+
+      return $values;
+  }
+
+  /**
+   * Update the table with the existing data
+   *
+   * @param array $data
+   * @return void
+   */
+  public function processData($data) {
+    if(empty($data)) {
+      return;
+    }
+
+
+//    // The users view is no longer required.
+    // XXX Each time i pull the data from the operations portal all of them will be udpated because
+    //     at least the column last_updated will have a new value. Perhaps the best approach is to
+    // Create temp table does not inherit indexes and still we need to copy. Perhaps the best approach is
+    // to clone the table, delete the old one and then rename the temp table
+    $db = ConnectionManager::getDataSource('default');
+
+    if(isset($db->config['prefix'])) {
+      $prefix = $db->config['prefix'];
+    }
+
+    // Create table as cm_voms_members only schema
+    $table = ($prefix ?? '') . Inflector::tableize($this->name);
+    try {
+      $this->VomsMemberClone = ClassRegistry::init('VoMember.VomsMemberClone');
+      $this->VomsMemberClone->tblCreate($table);
+      // save the data
+      $this->VomsMemberClone->importData($data);
+
+      $db->begin();
+      // Drop the current table
+      $this->query('DROP TABLE IF EXISTS ' . $table . ' CASCADE');
+      // Rename the tmp table
+      $this->VomsMemberClone->tableRename($table);
+    }
+    catch(Exception $e) {
+      $err = filter_var($e->getMessage(),FILTER_SANITIZE_SPECIAL_CHARS);
+      $this->log(__METHOD__ . "::error message => " . $err, LOG_DEBUG);
+      $db->rollback();
+      return;
+    }
+
+    $db->cacheSources = true;
```

Comment:
these could be inside `try` block? Or `db->begin()` should be outside of the block?

Reply:
cacheSources is unrelated to the transaction but needs to happen just before we  commit since we have to make sure the the transaction actions succeeded. Also the functions above the begin have their own transaction handling internally.